### PR TITLE
Fix delta eds bug: incremental eds will only generate eds for the upd…

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
…ated services so if we cal removed resources for incremental eds, it will cause missing endpoints in envoy. We should not cal removed resource when it is incremental eds push

**Please provide a description of this PR:**

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/xds/delta.go [hunk 1]
```diff
@@ -469,21 +468,23 @@ func (s *DiscoveryServer) pushDeltaXds(con *Connection, push *model.PushContext,
 		Nonce:             nonce(push.LedgerVersion),
 		Resources:         res,
 	}
+
+	currentResources := extractNames(res)
 	if usedDelta {
 		resp.RemovedResources = deletedRes
-	} else {
+	} else if req.Full {
 		// similar to sotw
```
Comment (howardjohn) [inline pilot/pkg/xds/delta.go:476]: add some comments maybe?

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/xds/delta.go [hunk 2]
```diff
@@ -446,7 +446,6 @@ func (s *DiscoveryServer) pushDeltaXds(con *Connection, push *model.PushContext,
 	}
 	defer func() { recordPushTime(w.TypeUrl, time.Since(t0)) }()
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/xds/delta.go:448]: oh, where does envoy say that?

File: pilot/pkg/xds/delta.go [hunk 3]
```diff
@@ -469,21 +468,23 @@ func (s *DiscoveryServer) pushDeltaXds(con *Connection, push *model.PushContext,
 		Nonce:             nonce(push.LedgerVersion),
 		Resources:         res,
 	}
+
+	currentResources := extractNames(res)
 	if usedDelta {
 		resp.RemovedResources = deletedRes
-	} else {
+	} else if req.Full {
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/xds/delta.go:475]: i think we can mock it

**Comment by howardjohn** [comment]
'When the DeltaDiscoveryRequest is a ACK or NACK message in response to a
previous DeltaDiscoveryResponse, the response_nonce must be the nonce in
the DeltaDiscoveryResponse. Otherwise (unlike in DiscoveryRequest)
response_nonce must be omitted.'

In the
https://www.envoyproxy.io/docs/envoy/latest/api-v3/service/discovery/v3/discovery.proto#envoy-v3-api-msg-service-discovery-v3-deltadiscoveryrequest

**Comment by hzxuzhonghu** [comment]
Thanks

**Comment by hzxuzhonghu** [comment]
It is not reflecting `Envoy will ignore RemoveResources in EDS`

**Pull request merged**